File: util/modif_trajet.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>

static const char *ENTREE = "trajets-c.bin";

typedef char Enregistrement[10];

#define N 838860

int
verifier (const char *entree)
{
  int _retour = 1;
  FILE *_flux = fopen (entree, "rb");
  long _n = 0;
  size_t _i = 0;
  Enregistrement *_enregistrements = malloc (N * sizeof (Enregistrement));
  size_t _nb_lus = 0;

  while (1)
    {
      _nb_lus = fread (_enregistrements, sizeof (Enregistrement), N, _flux);
      if (_nb_lus == 0)
        {
          fprintf (stderr, "Fin de vérification\n");
          fflush (stderr);
          break;
        }
      for (_i = 0; _i < _nb_lus; ++_i)
        {
          if (_enregistrements[_i][6] != 0)
            {
              fprintf (stderr, "Octet utile ! %u\n", _enregistrements[_i][6]);
              fflush (stderr);
              _retour = 0;
              break;
            }
          _n += 1;
          if (_n % 10000000 == 0)
            {
              fprintf (stderr, "%zuM enregistrements vérifiés\n",
                       _n / 1000000);
              fflush (stderr);
            }
        }
    }

  fclose (_flux);
  free (_enregistrements);
  return _retour;
}


void
modifier (const char *chemin)
{
  FILE *_entree = fopen (chemin, "rb");
  FILE *_sortie = fopen (chemin, "r+b");
  Enregistrement _enregistrement;
  size_t _nb_lus = 0;
  long _n = 0;
  size_t _taille = 0;

  fseek (_entree, 7, SEEK_SET);
  fseek (_sortie, 6, SEEK_SET);

  // Parties conservées :
  // - fin de l'enregistrement N
  // - début de l'enregistrement N+1
  while (1)
    {
      _nb_lus = fread (_enregistrement, sizeof (Enregistrement), 1, _entree);
      if (_nb_lus != 1)
        {
          break;
        }
      fwrite (_enregistrement, 9, 1, _sortie);

      _n += 1;
      if (_n % 10000000 == 0)
        {
          fprintf (stderr, "%zuM enregistrements modifiés\n", _n / 1000000);
          fflush (stderr);
        }
    }

  // Fin du dernier enregistrement
  fread (_enregistrement, 3, 1, _entree);
  fwrite (_enregistrement, 3, 1, _sortie);

  // Troncature
  _taille = ftell (_sortie);
  fclose (_sortie);
  fclose (_entree);
  truncate (chemin, _taille);
}


int
main (void)
{
  if (verifier (ENTREE))
    {
      modifier (ENTREE);
    }

  return 0;
}

File: util/modif_trajet.py
#!/usr/bin/python3 -O
# -*- coding: utf-8 -*-

"""Supprime un octet inutile du fichier de trajets
"""

import logging

ENTRÉE = "trajets-c.bin"


def vérifier(entrée):
    """Vrai ssi le 7ème octet de chaque enregistrement de 10 octets vaut 0
    """
    retour = True
    with open(entrée, "rb") as flux:
        flux.seek(0, 2)
        taille = flux.tell()
        if taille % 10 != 0:
            logging.error("Taille incorrecte : {} octets".format(taille))
            retour = False
        else:
            logging.info("Taille correcte : {} octets".format(taille))
        flux.seek(0, 0)

        n = 0
        while retour:
            données = flux.read(10)
            if len(données) == 0:
                logging.info("Fin de vérification")
                break
            if données[6] != 0:
                logging.error("Octet utile ! {}".format(flux.tell() - 10))
                retour = False
            n += 1
            if n % 10000000 == 0:
                logging.info(
                    "{}M enregistrements vérifiés".format(n // 1000000))
    return retour


def modifier(entrée):
    """Supprime le 7ème octet de chaque enregistrement de 10 octets
    """
    é = 6
    l = 7
    n = 0
    # il manque volontairement un '+' le temps de la mise-au-point
    with open(entrée, "rb") as flux:
        flux.seek(0, 2)
        taille = flux.tell() - 3

        # Partie conservée : fin de l'enregistrement N, début de
        # l'enregistrement N+1
        while l < taille:
            flux.seek(l, 0)
            données = flux.read(9)
            l += 10

            flux.seek(é, 0)
            flux.write(données)
            é += 9

            n += 1
            if n % 1000000 == 0:
                logging.info(
                    "{}M enregistrements traités".format(n // 1000000))

        # Fin du dernier enregistrement
        flux.seek(l, 0)
        données = flux.read(3)

        flux.seek(é, 0)
        flux.write(données)
        é += 3

        # Troncature
        flux.truncate(é)


if __name__ == "__main__":
    logging.basicConfig(datefmt="%d/%m %Hh%M",
                        format="%(asctime)s | %(message)s",
                        level=logging.INFO)

    if vérifier(ENTRÉE):
        modifier(ENTRÉE)
